refactor(rome_formatter): implement snapshot testing with insta.rs

Boshen opened the pull request:
## Summary
### Rome Formatter
This PR demonstrates snapshot testing using https://github.com/mitsuhiko/insta

This aids faster TDD cycle where we don't have to manually change the expected output.

The code setup mimics rome-classic, where the input and snapshot files are put side by side, instead of putting all snapshots inside a single directory.

I can migrate the current tests if we prefer this approach.

## Test Plan
First we set up the snapshot by just running the test:
```
cargo test -p rome_formatter snapshot_test
```

Then we get a colored diff:

<img src="https://user-images.githubusercontent.com/1430279/145156256-b9c1bda5-0af3-47e4-822a-b2213900f112.png" width="600">

Or with companion test tool `cargo insta test`
<img src="https://user-images.githubusercontent.com/1430279/145168944-8a112304-73b4-481b-9a6e-3d517a829fdb.png" width="600">

Finally with companion review tool `cargo install cargo-insta`, we run `cargo insta review`

<img src="https://user-images.githubusercontent.com/1430279/145156454-d3e8c66f-7914-48d8-a6aa-9f5bad5bb9a3.png" width="400">

accept or reject, and then we are done.

MichaReiser commented:
This looks nice! I wonder what we should do about the parser tests that currently use `expected`. I would rather only use a single solution for snapshot tests

https://github.com/rome/tools/blob/10d4a0be9cfe7f42b5da15e3d32f700b41172003/crates/rslint_parser/src/tests.rs#L132

MichaReiser commented on crates/rome_formatter/tests/snapshot_test.rs:
```diff
@@ -0,0 +1,20 @@
+use rome_formatter::{format_file, FormatOptions};
+
+#[test]
+fn snapshot_test() {
```
Is this now alongside the existing spec tests? Does this generate a single test for all files?

What I liked about our existing snapshot tests is that you can run all tests at the same time and see all failing tests (and not just the first file that has a snapshot mismatch)

Boshen replied:
Good catch, I found `cargo insta test` where it runs everything. (Test plan updated)

MichaReiser replied:
Can we integrate this into our existing test infra rather than creating something separate? 

https://github.com/rome/tools/blob/a1c49f0e06826be8cb95f25b4b3e742d9a363739/crates/rome_formatter/tests/spec_test.rs

It would mainly mean to just move this code into the `spec_test` function.

Boshen commented:
@MichaReiser This PR is for demonstration, 

> This looks nice! I wonder what we should do about the parser tests that currently use `expected`. I would rather only use a single solution for snapshot tests
> 
> https://github.com/rome/tools/blob/10d4a0be9cfe7f42b5da15e3d32f700b41172003/crates/rslint_parser/src/tests.rs#L132

This PR is for demonstration. I'll migrate the exiting tests if we are good with insta.rs

ematipico commented on crates/rome_formatter/tests/specs/js/continue_stmt.js.snap:
```diff
@@ -0,0 +1,13 @@
+---
```
One thing that I liked of our old snapshot infrastructure (the old Rome in TypeScript) it that we could also **append the original** source code in the snapshot. 

Is it something that we can do with `insta.rs`?

Boshen replied:
I prepended the input snippet, now it looks this:
```
---
source: crates/rome_formatter/tests/snapshot_test.rs
expression: arg_list.js
input_file: crates/rome_formatter/tests/specs/js/arg_list.js

---
# Input
function foo(someotherlongvariableshould, someotherlongvariableshould, someotherlongvariableshould, someotherlongvariableshould, someotherlongvariableshould, someotherlongvariableshould, someotherlongvariableshould) {
	return "nothing";
}

# Output
function foo(
	someotherlongvariableshould,
	someotherlongvariableshould,
	someotherlongvariableshould,
	someotherlongvariableshould,
	someotherlongvariableshould,
	someotherlongvariableshould,
	someotherlongvariableshould
) {
	return "nothing";
}
```

MichaReiser commented on crates/rome_formatter/tests/snapshot_test.rs:
```diff
@@ -0,0 +1,22 @@
+
+	insta::glob!("specs/js/**/*.js", |file_path| {
+		let mut settings = insta::Settings::clone_current();
+		settings.set_sort_maps(true);
+		settings.set_prepend_module_to_snapshot(false);
+		settings.set_snapshot_path(file_path.parent().unwrap());
+		settings.set_snapshot_suffix("");
+		settings.bind(|| {
```
Do we need all these settings? 

For example: `sort_maps`: We're snapshoting strings, so there are no maps to sort. 

You can also use the `with_settings` macro https://docs.rs/insta/latest/insta/macro.with_settings.html

Boshen replied:
The settings are for getting the minimal file name output and putting it in the right directory, I can change it to the macro but they are the same thing.

xunilrj commented:
The experience when a test fail in the CI is a little bit confusing. Not entirely sure what failed. For example:

```
---- snapshot_test stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Differences ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot file: crates/rome_formatter/tests/specs/js/binary.js.snap
Snapshot: binary.js
Source: crates/rome_formatter/tests/snapshot_test.rs:19
Input file: crates/rome_formatter/tests/specs/js/binary.js
+new results
────────────────────────────────────────────────────────────────────────────────
binary.js
────────────┬───────────────────────────────────────────────────────────────────
          0 │+# Input
          1 │+a  +   b
          2 │+a  <   b
          3 │+a  >   b
          4 │+a  <=  b
          5 │+a  >=  b
          6 │+a  ==  b
          7 │+a  === b
          8 │+a  !=  b
          9 │+a  !== b
         10 │+a  +   b
         11 │+a  -   b
         12 │+a  *   b
         13 │+a  /   b
         14 │+a  %   b
         15 │+a  **  b
         16 │+a  <<  b
         17 │+a  >>  b
         18 │+a  >>> b
         19 │+a  &   b
         20 │+a  |   b
         21 │+a  ^   b
         22 │+a  in  b
         23 │+a  instanceof b
         24 │+
         25 │+# Output
         26 │+a + b;
         27 │+a < b;
         28 │+a > b;
         29 │+a <= b;
         30 │+a >= b;
         31 │+a == b;
         32 │+a === b;
         33 │+a != b;
         34 │+a !== b;
         35 │+a + b;
         36 │+a - b;
         37 │+a * b;
         38 │+a / b;
         39 │+a % b;
         40 │+a ** b;
         41 │+a << b;
         42 │+a >> b;
         43 │+a >>> b;
         44 │+a & b;
         45 │+a | b;
         46 │+a ^ b;
         47 │+a in b;
         48 │+a instanceof b;
────────────┴───────────────────────────────────────────────────────────────────
thread 'main' panicked at 'snapshot assertion for 'binary.js' failed in line 19', 
```

ematipico commented:
@xunilrj  what would you expect instead? This would help to understand what we're looking for

Boshen commented:
> The experience when a test fail in the CI is a little bit confusing. Not entirely sure what failed. For example:

This example is a bit confusing due to it's a entirely new results, hence the little
```
+new results
```
in the header.

It'll be less confusing once we get a smaller diff.

And in my opinion, the diff is better than the current failure output, e.g.
```
---- formatter::js::switch_statement stdout ----
thread 'formatter::js::switch_statement' panicked at 'assertion failed: `(left == right)`
  left: `"switch (key) {\n\tcase value:\n\t\tbreak\n\tdefault:\n\t\tbreak;\n}\n"`,
 right: `"switch (key) {\n\tcase value:\n\t\tbreak;\n\tdefault:\n\t\tbreak;\n}\n"`', crates/rome_formatter/tests/spec_test.rs:38:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

where we have to squeeze our eyes to find the diff :-/

xunilrj commented:
hum... I thought it was something with the ``` 19 │+a  &   b``` because of

```
thread 'main' panicked at 'snapshot assertion for 'binary.js' failed in line 19', 
```

> @xunilrj what would you expect instead? This would help to understand what we're looking for

Just something like "this line here is different". But I got confused because this whole test is new.

Boshen commented:
What's the consensus here? I'll clean this up and migrate the existing tests if we are good with `insta.rs`, or we can close this.

ematipico commented:
I think we should switch to it. Eventually Rome will have its own solution but for the time being I'd like to move to something better like `insta`.

We can personalize the content of the snapshots too, so it's really nice! And there's a VSCode extension too! We can't ask for more :)

xunilrj commented:
> I think we should switch to it. Eventually Rome will have its own solution but for the time being I'd like to move to something better like `insta`.
> 
> We can personalize the content of the snapshots too, so it's really nice! And there's a VSCode extension too! We can't ask for more :)

+1

ematipico commented:
@yassere  what do you think?

yassere commented:
> @yassere what do you think?

I've never used `insta` but those screen shots look nice. Does anyone know why `rust-analyzer` switched from `insta` to `expect-test` and if any of those reasons are still applicable?

MichaReiser commented:
> > @yassere what do you think?
> 
> I've never used `insta` but those screen shots look nice. Does anyone know why `rust-analyzer` switched from `insta` to `expect-test` and if any of those reasons are still applicable?

I believe they used `expect` from the beginning? 

https://github.com/rust-analyzer/rust-analyzer/commit/b0fd3faf36c94c3fc52151c6aa82b36b43b7cceb#diff-b0bdbb712b95134e052c53a51bc26c9ef2e05e1ab25ae88126949b1a47f456f9

yassere commented:
> I believe they used `expect` from the beginning?

I found PR `rust-analyzer/pull/5101` (I don't want to create a backlink) where they discussed creating an alternative to `insta` and it links to an issue with some more discussion. I haven't gone through it all, so I don't know if any of it is relevant to us (or if `insta` itself has changed in the meantime).

yassere commented:
To clarify, I just wanted to link that PR for context, but rust-analyzer's needs don't necessarily match our own. I have no objection to using `insta` if that's the consensus. It looks quite nice.

Boshen commented:
All done and ready to review, the steps I look I took are separated into commits: 
1d67679 test(formatter): add insta.rs snapshot testing
90782d1 chore(formatter): remove old {js,json}/**/*.expect.{js,json} to prepare for insta tests
6bdd9df chore(formatter): remove old spec test setup
6c2818a test(formatter): generate new insta.rs snapshots

MichaReiser commented:
Nice. Would you mind updating the contribution document so that people know how to run tests and update the snapshots?

MichaReiser commented on crates/rome_formatter/tests/snapshot_tests.rs:
```diff
@@ -0,0 +1,42 @@
+	}
+
+	fn snapshot_test(glob: &str) {
+		insta::glob!(glob, |file_path| {
```
Would it be possible to still use our spec tests or is there a downside to it? I liked it that tests can run in parallel and that you see all failing tests by name using cargo test (which has IDE integration)

Boshen replied:
It's possible, I'll take a look at the original macro and remove this glob

Boshen commented:
@MichaReiser updated with https://github.com/rome/tools/pull/1857/commits/489975186011c0c6e3cd943786f18ed06183b923 where we only have to iterate the input file, everything is much better now 😄

ematipico reviewed (changes requested):
It would be great if we created a new  section inside the `CONTRIBUTING.md` and explains how to write tests for the formatter - watch them and update them. 

We already have a section for the parser, we could do the same for the formatter. So sorry to request changes, but I would like to cover this section right now, until the knowledge is fresh.

Boshen commented:
I wrote a small guide on how to TDD with insta and also how to find the corresponding implementation for a syntax: https://github.com/rome/tools/pull/1857/commits/2bf313920d09906a0ef53fe9206b056fe32e398f

Feel free to edit and update its contents.

ematipico commented on CONTRIBUTING.md:
````diff
@@ -180,3 +180,47 @@ The command will tell the test suite to generate and update the `.rast` files.
+cargo watch -i '*.new' -x 'test -p rome_formatter formatter'
+```
+
+To initialize a new snapshot for javascript, add a new input file to `crates/rome_formatter/tests/specs/javascript`, e.g. `arrow.js`
````
```suggestion
To initialize a new snapshot for javascript, add a new input file in the directory `crates/rome_formatter/tests/specs/javascript`, e.g. `arrow.js`
```
```suggestion
To create a new snapshot test for JavaScript, create a new file to `crates/rome_formatter/tests/specs/javascript`, e.g. `arrow.js`
```

ematipico commented on CONTRIBUTING.md:
```diff
@@ -180,3 +180,47 @@ The command will tell the test suite to generate and update the `.rast` files.
+
+#### Finding the corresponding formatter implementation
+
+To find the corresponding implementation for the arrow function above, we can start from logging the parsed syntax tree in `crates/rome_formatter/src/lib.rs`:
```
Is this still related to testing? It seems that were we are talking about the actual implementation and debugging of the formatter.

If so, I would remove it from the PR. We can add a new chapter about "implementing something for the formatter" in another PR

Boshen commented:
@ematipico I've clean it up in the last commit, https://github.com/rome/tools/blob/6dcce274b85b0f92c4753a9fc890a84958137b23/CONTRIBUTING.md#write-snapshot-tests-for-the-formatter

* added the basic `test` command to generate the test
* removed the "formatter implementation" part,

MichaReiser commented:
Looks good. You'll need to re-generate all snapshots before merging (and delete all old snapshots).

Boshen commented:
Rebased and regenerated all snapshots, ready to merge.
